io/blob.go

```diff
@@ -0,0 +1,311 @@
+	return 0, &fs.PathError{Op: "readAt", Path: d.key, Err: fs.ErrInvalid}
+}
+func (d *iofsDir) Seek(offset int64, whence int) (int64, error) {
+	return 0, &fs.PathError{Op: "seek", Path: d.key, Err: fs.ErrInvalid}
+}
+func (d *iofsDir) Close() error { return nil }
+func (d *iofsDir) ReadDir(count int) ([]fs.DirEntry, error) {
+	if err := d.openOnce(); err != nil {
+		return nil, err
+	}
+	n := len(d.entries) - d.offset
+	if n == 0 && count > 0 {
+		return nil, io.EOF
+	}
+	if count > 0 && n > count {
+		n = count
+	}
+	list := make([]fs.DirEntry, n)
+	for i := range list {
+		list[i] = d.entries[d.offset+i]
+	}
+	d.offset += n
+	return list, nil
+}
+
+func (d *iofsDir) openOnce() error {
+	if d.opened {
+		return nil
+	}
+	d.opened = true
+
+	// blob expects directories to end in the delimiter, except at the top level.
+	prefix := d.key
+	if prefix != "" {
+		prefix += "/"
+	}
+	listOpts := blob.ListOptions{
+		Prefix:    prefix,
+		Delimiter: "/",
+	}
+	ctx := d.b.ctx
+
+	// Fetch all the directory entries.
+	// Conceivably we could only fetch a few here, and fetch the rest lazily
+	// on demand, but that would add significant complexity.
+	iter := d.b.List(&listOpts)
+	for {
+		item, err := iter.Next(ctx)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		name := filepath.Base(item.Key)
+		if item.IsDir {
+			d.entries = append(d.entries, newDir(d.b, item.Key, name))
+		} else {
+			d.entries = append(d.entries, &iofsFileInfo{item, name})
+		}
+	}
+	// There is no such thing as an empty directory in Bucket, so if
+	// we didn't find anything, it doesn't exist.
+	if len(d.entries) == 0 {
+		return fs.ErrNotExist
+	}
+	return nil
+}
+
+// blobOpenFile describes a single open blob as a File.
+// It implements the iceberg-go/io.File interface.
+// It is based on gocloud.dev/blob.iofsOpenFile which:
+// - Doesn't support the `io.ReaderAt` interface
+// - Is not externally accessible, so copied here
+type blobOpenFile struct {
+	*blob.Reader
+	name string
+}
+
+func (f *blobOpenFile) ReadAt(p []byte, off int64) (int, error) {
+	finalOff, err := f.Reader.Seek(off, io.SeekStart)
+	if err != nil {
+		return -1, err
+	} else if finalOff != off {
+		return -1, io.ErrUnexpectedEOF
+	}
+
+	return f.Read(p)
+}
+
+// Functions to implement the `Stat()` function in the `io/fs.File` interface
+
+func (f *blobOpenFile) Name() string               { return f.name }
+func (f *blobOpenFile) Mode() fs.FileMode          { return fs.ModeIrregular }
+func (f *blobOpenFile) Sys() interface{}           { return f.Reader }
+func (f *blobOpenFile) IsDir() bool                { return false }
+func (f *blobOpenFile) Stat() (fs.FileInfo, error) { return f, nil }
+
+// BlobFileIO represents a file system backed by a bucket in object store. It implements the `iceberg-go/io.FileIO` interface.
+type BlobFileIO struct {
+	*blob.Bucket
+	ctx    context.Context
+	opts   *blob.ReaderOptions
+	prefix string
+}

```

Comment:
I tend to be more conservative in what we actually export. Is there any need to export this type as opposed to just let it be used through the interfaces?

Reply:
Changed to unexported